File: src/rac/services/gate.py
"""Policy-aware unified enforcement — ``rac gate`` (v0.21.14, ADR-049 / ADR-063).

``rac gate`` is the single enforcement entry point. It runs validation,
relationship integrity, and review over one corpus, normalises every finding to
a :class:`GateFinding`, then classifies each as *blocking* or *advisory* under
the corpus enforcement policy (ADR-049 — enforcement is the product). The result
is one exit code, one JSON envelope, and one SARIF document, so a PR gate carries
the whole RAC contract as a single required check instead of three uploads.

Classification is governed, not hardcoded (ADR-063: the thin client renders, the
engine decides). A corpus declares an optional ``enforcement:`` section in its
committed ``.rac/config.yaml`` (owned and loaded by :mod:`rac.services.init`)
mapping finding codes to ``blocking`` / ``advisory`` / ``off``. The *default*
enforcement classes are chosen so that, with no policy, a gate run is ``ok``
exactly when validate, relationships, and review all pass — the v0.21.13
behaviour the policy refines.

Deterministic and offline (ADR-002): the gate composes the same deterministic
services, applies a pure policy pass, and sorts findings by a stable key, so an
unchanged corpus yields byte-identical report, JSON, and SARIF.

Single corpus walk (v0.21.19): the three analyses share one pre-walked snapshot
via the engine's ``*_from_corpus`` seams instead of each re-walking the tree — one
walk, not three — with output byte-identical to the prior multi-walk composition
(ADR-023). ``test_gate_perf`` pins that equivalence, and imports the private
finding helpers below to reconstruct the report the multi-walk way; their names,
signatures, and ``(code, severity, path, line, message)`` tuple shape are a hard
contract.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from rac.core.corpus import walk_corpus
from rac.services.init import load_enforcement_policy, load_overrides
from rac.services.portfolio import portfolio_from_corpus
from rac.services.relationships import (
    RELATIONSHIP_SEVERITY,
    RelationshipIssue,
    validation_from_corpus,
)
from rac.services.review import (
    PRIORITY_BROKEN_RELATIONSHIP,
    ReviewIssue,
    review_from_portfolio,
)
from rac.services.validate import DirectoryValidation, validate_corpus

# The two enforcement classes a finding can carry, plus the suppressed marker
# (``off`` drops a finding entirely). Sources name where a finding originated.
ENFORCEMENT_BLOCKING = "blocking"
ENFORCEMENT_ADVISORY = "advisory"
ENFORCEMENT_OFF = "off"

SOURCE_VALIDATE = "validate"
SOURCE_RELATIONSHIPS = "relationships"
SOURCE_REVIEW = "review"


@dataclass(frozen=True)
class EnforcementPolicy:
    """A corpus enforcement policy: finding codes mapped to a class (ADR-049).

    Three disjoint *intent* sets. :meth:`classify` resolves a finding's effective
    class with a fixed precedence — ``off`` (suppress) wins, then ``blocking``,
    then ``advisory``, else the caller's default. A code declared in more than one
    set is therefore harmless: precedence makes the result independent of
    declaration order.
    """

    blocking: frozenset[str] = frozenset()
    advisory: frozenset[str] = frozenset()
    off: frozenset[str] = frozenset()

    def classify(self, code: str, default: str) -> str | None:
        """Effective enforcement class for ``code``, or ``None`` when suppressed.

        Precedence ``off`` -> ``blocking`` -> ``advisory`` -> ``default``. A
        ``None`` result means the policy turned this finding off and it is dropped.
        """
        if code in self.off:
            return None
        if code in self.blocking:
            return ENFORCEMENT_BLOCKING
        if code in self.advisory:
            return ENFORCEMENT_ADVISORY
        return default


# The neutral policy: every finding keeps its default class. Used when a corpus
# declares no ``enforcement:`` section, making the no-policy path a pure no-op.
EMPTY_POLICY = EnforcementPolicy()

# Default enforcement for a validate finding, keyed by intrinsic severity. Only an
# ``error`` (which sets the invalid status) is blocking; warnings and OKF info
# findings fall through to advisory. Imported by ``test_gate_perf``.
_VALIDATE_DEFAULT = {"error": ENFORCEMENT_BLOCKING}


@dataclass(frozen=True)
class GateFinding:
    """One enforced finding, normalised across the three underlying services.

    ``source`` records which service produced it; ``severity`` is the intrinsic
    severity (drives the SARIF level); ``enforcement`` is the policy-resolved class
    that drives the exit code. ``to_dict`` is the stable JSON contract (ADR-007),
    ordered for deterministic output.
    """

    source: str  # SOURCE_VALIDATE | SOURCE_RELATIONSHIPS | SOURCE_REVIEW
    code: str
    severity: str  # "error" | "warning" | "info"
    enforcement: str  # ENFORCEMENT_BLOCKING | ENFORCEMENT_ADVISORY
    path: str
    line: int | None
    message: str

    def to_dict(self) -> dict:
        return {
            "source": self.source,
            "code": self.code,
            "severity": self.severity,
            "enforcement": self.enforcement,
            "path": self.path,
            "line": self.line,
            "message": self.message,
        }


@dataclass
class GateReport:
    """The unified enforcement result over a corpus (v0.21.14).

    ``ok`` is False when any finding is blocking — the single exit-code signal the
    PR gate consumes. ``to_dict`` is the stable JSON contract (ADR-007).
    """

    directory: str
    recursive: bool
    findings: list[GateFinding] = field(default_factory=list)

    @property
    def blocking(self) -> list[GateFinding]:
        return [f for f in self.findings if f.enforcement == ENFORCEMENT_BLOCKING]

    @property
    def advisory(self) -> list[GateFinding]:
        return [f for f in self.findings if f.enforcement == ENFORCEMENT_ADVISORY]

    @property
    def ok(self) -> bool:
        """True when nothing is blocking — advisory findings never fail the gate."""
        return not self.blocking

    def to_dict(self) -> dict:
        return {
            "schema_version": "1",
            "directory": self.directory,
            "recursive": self.recursive,
            "ok": self.ok,
            "blocking_count": len(self.blocking),
            "advisory_count": len(self.advisory),
            "findings": [f.to_dict() for f in self.findings],
        }


# --- Per-source finding normalisers ------------------------------------------
# Each collapses a service's finding into the gate's common
# ``(code, severity, path, line, message)`` tuple. The tuple shape and these three
# names are imported by ``test_gate_perf`` — a hard contract.


def _validate_findings(result: DirectoryValidation) -> list[tuple[str, str, str, int | None, str]]:
    """``(code, severity, path, line, message)`` for every validate finding.

    Core validation issues carry a line anchor when present; OKF conformance
    findings are file-level (no line). Mirrors ``render_validate_sarif``.
    """
    out: list[tuple[str, str, str, int | None, str]] = []
    for file in result.files:
        for issue in file.issues:
            out.append((issue.code, issue.severity, file.path, issue.line, issue.message))
    if result.okf is not None:
        for finding in result.okf.findings:
            out.append((finding.code, finding.severity, finding.path, None, finding.message))
    return out


def _relationship_finding(issue: RelationshipIssue) -> tuple[str, str, str, int | None, str]:
    """``(code, severity, path, line, message)`` for one relationship finding.

    Message and anchor are borrowed from the SARIF result builder so ``rac gate``
    and ``rac relationships --sarif`` share one formatting source and can never
    drift. Line is always None (relationship findings are file-level).
    """
    # Imported lazily: rac.output.sarif imports GateReport from this module, so a
    # module-level import here would close a cycle.
    from rac.output.sarif import _relationship_result

    result = _relationship_result(issue)
    uri = result["locations"][0]["physicalLocation"]["artifactLocation"]["uri"]
    severity = RELATIONSHIP_SEVERITY.get(issue.code, "warning")
    return (issue.code, severity, uri, None, result["message"]["text"])


def _review_finding(issue: ReviewIssue) -> tuple[str, str, str, int | None, str]:
    """``(code, severity, path, line, message)`` for one review finding.

    The suggested action is appended so the fix is visible inline, exactly as
    ``render_review_sarif`` formats it. Line is always None.
    """
    message = f"{issue.message} — {issue.action}" if issue.action else issue.message
    return (issue.code, issue.severity, issue.path, None, message)


def build_gate(
    directory: str,
    recursive: bool = True,
    policy: EnforcementPolicy | None = None,
) -> GateReport:
    """Run validation, relationships, and review and enforce the corpus policy.

    When ``policy`` is None it is loaded from ``.rac/config.yaml``
    (:func:`load_enforcement_policy`). Each underlying finding is normalised, given
    a default enforcement class, then run through the policy — findings the policy
    turns ``off`` are dropped. Findings are sorted by
    ``(path, line, source, code, message)`` so report, JSON, and SARIF are
    byte-stable (ADR-002).

    One walk feeds every analysis (v0.21.19): the ``*_from_corpus`` snapshot seams
    produce results identical to their directory entry points
    (``validate_directory`` / ``validate_relationships`` / ``build_review``), which
    each walk once. Severity overrides (ADR-053) are loaded once and applied
    identically by every snapshot path.
    """
    if policy is None:
        policy = load_enforcement_policy(directory)

    entries = list(walk_corpus(directory, recursive=recursive))
    overrides = load_overrides(directory)

    validation = validate_corpus(directory, entries, recursive=recursive, overrides=overrides)
    relationships = validation_from_corpus(directory, entries, recursive=recursive)
    portfolio = portfolio_from_corpus(directory, entries, recursive=recursive)
    review = review_from_portfolio(directory, portfolio, recursive=recursive)

    findings: list[GateFinding] = []

    def enforce(
        source: str,
        code: str,
        severity: str,
        path: str,
        line: int | None,
        message: str,
        default: str,
    ) -> None:
        enforcement = policy.classify(code, default)
        if enforcement is None:
            return  # suppressed by an ``off`` policy entry
        findings.append(
            GateFinding(
                source=source,
                code=code,
                severity=severity,
                enforcement=enforcement,
                path=path,
                line=line,
                message=message,
            )
        )

    # Validate: only an intrinsic error is blocking by default (it is what sets the
    # invalid status); warnings and OKF info findings default to advisory.
    for code, severity, path, line, message in _validate_findings(validation):
        enforce(
            SOURCE_VALIDATE,
            code,
            severity,
            path,
            line,
            message,
            _VALIDATE_DEFAULT.get(severity, ENFORCEMENT_ADVISORY),
        )

    # Relationships: every issue fails ``--validate`` today, so each defaults to
    # blocking; a policy may downgrade a specific code (e.g. superseded).
    for rel_issue in relationships.issues:
        code, severity, path, line, message = _relationship_finding(rel_issue)
        enforce(SOURCE_RELATIONSHIPS, code, severity, path, line, message, ENFORCEMENT_BLOCKING)

    # Review: priority 1-2 findings fail review today (ReviewReport.ok), so they
    # default to blocking; advisory priorities (3+) default to advisory.
    for review_issue in review.issues:
        code, severity, path, line, message = _review_finding(review_issue)
        default = (
            ENFORCEMENT_BLOCKING
            if review_issue.priority <= PRIORITY_BROKEN_RELATIONSHIP
            else ENFORCEMENT_ADVISORY
        )
        enforce(SOURCE_REVIEW, code, severity, path, line, message, default)

    findings.sort(key=lambda f: (f.path, f.line or 0, f.source, f.code, f.message))
    return GateReport(directory=directory, recursive=recursive, findings=findings)

File: rust/rac-engine/src/gate.rs
//! Policy-aware unified enforcement (`decided.services.gate`, v0.21.14 /
//! ADR-049) plus the STRICT `.decided/config.yaml` loaders it alone consumes
//! (`decided.services.init.load_enforcement_policy` / `load_overrides`, the
//! raising paths).
//!
//! `decided gate` composes validation, relationship integrity, and review over
//! one corpus, then classifies every finding as blocking or advisory under
//! the corpus enforcement policy. The other commands keep the engine's
//! lenient config readers (`validate::load_overrides` skips malformed
//! entries); the gate is the one surface where a malformed config is an
//! operational error — `decided: malformed repository config <path>: <reason>`
//! on stderr, exit 1 (`MalformedRepositoryConfig` in `cmd_gate`).

use crate::commands::{validate_directory, DirectoryValidation};
use crate::frontmatter::{yaml_load_config, Yaml};
use crate::output::relationship_sarif_parts;
use crate::portfolio::portfolio_from_corpus;
use crate::relationships::{
    corpus_items, relationship_severity, validate_relationships, RelationshipValidation,
};
use crate::review::{review_from_portfolio, ReviewReport, PRIORITY_BROKEN_RELATIONSHIP};
use crate::validate::find_config_file;

pub const ENFORCEMENT_BLOCKING: &str = "blocking";
pub const ENFORCEMENT_ADVISORY: &str = "advisory";

pub const SOURCE_VALIDATE: &str = "validate";
pub const SOURCE_RELATIONSHIPS: &str = "relationships";
pub const SOURCE_REVIEW: &str = "review";

// ---------------------------------------------------------------------------
// MalformedRepositoryConfig (decided.services.init)
// ---------------------------------------------------------------------------

/// The gate's operational config error. `message()` is the oracle's
/// `str(MalformedRepositoryConfig)` shape; `cmd_gate` prefixes `decided: `.
#[derive(Debug)]
pub struct MalformedConfig {
    pub config_path: String,
    pub reason: String,
}

impl MalformedConfig {
    pub fn message(&self) -> String {
        format!(
            "malformed repository config {}: {}",
            self.config_path, self.reason
        )
    }
}

// ---------------------------------------------------------------------------
// EnforcementPolicy (decided.services.gate.EnforcementPolicy)
// ---------------------------------------------------------------------------

/// Finding codes mapped to an enforcement class (ADR-049). Precedence:
/// `off` (suppress, None) -> `blocking` -> `advisory` -> the caller default.
#[derive(Debug, Default)]
pub struct EnforcementPolicy {
    pub blocking: Vec<String>,
    pub advisory: Vec<String>,
    pub off: Vec<String>,
}

impl EnforcementPolicy {
    fn classify(&self, code: &str, default: &'static str) -> Option<&'static str> {
        if self.off.iter().any(|c| c == code) {
            return None;
        }
        if self.blocking.iter().any(|c| c == code) {
            return Some(ENFORCEMENT_BLOCKING);
        }
        if self.advisory.iter().any(|c| c == code) {
            return Some(ENFORCEMENT_ADVISORY);
        }
        Some(default)
    }
}

// ---------------------------------------------------------------------------
// Strict config loading (the raising `decided.services.init` paths)
// ---------------------------------------------------------------------------

/// `_parse_config_yaml`: read + YAML-parse the nearest config; a parse
/// failure is `invalid YAML: <problem>` (the oracle embeds PyYAML's exact
/// exception prose here — stderr is out of parity scope, so the engine's
/// own problem text stands in). Returns the top-level mapping pairs, or
/// None when the root is not a mapping (the loaders treat a non-mapping
/// root as "no section", matching `data.get(...) if isinstance(data, dict)`).
fn parse_config_pairs(
    config_path: &std::path::Path,
) -> Result<Option<Vec<(Yaml, Yaml)>>, MalformedConfig> {
    let display = config_path.to_string_lossy().into_owned();
    let text = std::fs::read_to_string(config_path).map_err(|e| MalformedConfig {
        config_path: display.clone(),
        reason: format!("invalid YAML: {e}"),
    })?;
    match yaml_load_config(&text) {
        Ok(Yaml::Map(pairs)) => Ok(Some(pairs)),
        Ok(_) => Ok(None),
        Err(problem) => Err(MalformedConfig {
            config_path: display,
            reason: format!("invalid YAML: {problem}"),
        }),
    }
}

fn yaml_get<'a>(pairs: &'a [(Yaml, Yaml)], name: &str) -> Option<&'a Yaml> {
    pairs.iter().find_map(|(k, v)| match k {
        Yaml::Str(s) if s == name => Some(v),
        _ => None,
    })
}

/// `_parse_code_list`: absent/null -> empty; anything but a list of strings
/// is a malformed config.
fn parse_code_list(
    config_path: &std::path::Path,
    value: Option<&Yaml>,
    where_: &str,
) -> Result<Vec<String>, MalformedConfig> {
    let malformed = || MalformedConfig {
        config_path: config_path.to_string_lossy().into_owned(),
        reason: format!("'{where_}' must be a list of finding-code strings"),
    };
    match value {
        None | Some(Yaml::Null) => Ok(Vec::new()),
        Some(Yaml::List(items)) => {
            let mut out = Vec::with_capacity(items.len());
            for item in items {
                match item {
                    Yaml::Str(s) => out.push(s.clone()),
                    _ => return Err(malformed()),
                }
            }
            Ok(out)
        }
        Some(_) => Err(malformed()),
    }
}

/// `load_enforcement_policy(start_dir)` — the STRICT reader `build_gate`
/// consumes. YAML 1.1 resolves a bare `off` key to Bool(false); both key
/// spellings are accepted, `off` winning (`section["off"] if "off" in
/// section else section.get(False)`).
pub fn load_enforcement_policy(start_dir: &str) -> Result<EnforcementPolicy, MalformedConfig> {
    let Some(config_path) = find_config_file(start_dir) else {
        return Ok(EnforcementPolicy::default());
    };
    let Some(pairs) = parse_config_pairs(&config_path)? else {
        return Ok(EnforcementPolicy::default());
    };
    let section = match yaml_get(&pairs, "enforcement") {
        None | Some(Yaml::Null) => return Ok(EnforcementPolicy::default()),
        Some(Yaml::Map(section)) => section,
        Some(_) => {
            return Err(MalformedConfig {
                config_path: config_path.to_string_lossy().into_owned(),
                reason: "'enforcement' must be a mapping".to_string(),
            })
        }
    };
    let blocking = parse_code_list(&config_path, yaml_get(section, "blocking"), "enforcement.blocking")?;
    let advisory = parse_code_list(&config_path, yaml_get(section, "advisory"), "enforcement.advisory")?;
    let off_value = yaml_get(section, "off").or_else(|| {
        section
            .iter()
            .find_map(|(k, v)| matches!(k, Yaml::Bool(false)).then_some(v))
    });
    let off = parse_code_list(&config_path, off_value, "enforcement.off")?;
    Ok(EnforcementPolicy {
        blocking,
        advisory,
        off,
    })
}

/// A Yaml key rendered the way Python's f-string would render the parsed
/// value inside `'validation.rules.<name>'` — best effort, stderr-only.
fn yaml_key_display(key: &Yaml) -> String {
    match key {
        Yaml::Str(s) => s.clone(),
        Yaml::Bool(true) => "True".to_string(),
        Yaml::Bool(false) => "False".to_string(),
        Yaml::Int(i) => i.to_string(),
        Yaml::Null => "None".to_string(),
        other => format!("{other:?}"),
    }
}

/// `_parse_severity_map` (strict): every entry must map a string name to an
/// allowed severity; YAML 1.1 bools coerce to `off`/`on` first.
fn check_severity_map(
    config_path: &std::path::Path,
    value: Option<&Yaml>,
    where_: &str,
    allowed: &[&str],
) -> Result<(), MalformedConfig> {
    let section = match value {
        None | Some(Yaml::Null) => return Ok(()),
        Some(Yaml::Map(section)) => section,
        Some(_) => {
            return Err(MalformedConfig {
                config_path: config_path.to_string_lossy().into_owned(),
                reason: format!("'{where_}' must be a mapping"),
            })
        }
    };
    for (name, sev) in section {
        let sev_text = match sev {
            Yaml::Bool(false) => Some("off".to_string()),
            Yaml::Bool(true) => Some("on".to_string()),
            Yaml::Str(s) => Some(s.clone()),
            _ => None,
        };
        let ok = matches!(name, Yaml::Str(_))
            && sev_text
                .as_deref()
                .map(|s| allowed.contains(&s))
                .unwrap_or(false);
        if !ok {
            return Err(MalformedConfig {
                config_path: config_path.to_string_lossy().into_owned(),
                reason: format!(
                    "'{where_}.{}' must map a name to one of {}",
                    yaml_key_display(name),
                    allowed.join(", ")
                ),
            });
        }
    }
    Ok(())
}

/// The STRICT face of `load_overrides` (ADR-053): `build_gate` validates the
/// `validation` section shape and raises where the oracle raises. The
/// override VALUES applied to the pipeline come from the engine's lenient
/// loader inside `validate_directory` — identical whenever this check
/// passes (the lenient reader only ever drops entries this one rejects).
pub fn check_overrides(start_dir: &str) -> Result<(), MalformedConfig> {
    let Some(config_path) = find_config_file(start_dir) else {
        return Ok(());
    };
    let Some(pairs) = parse_config_pairs(&config_path)? else {
        return Ok(());
    };
    let section = match yaml_get(&pairs, "validation") {
        None | Some(Yaml::Null) => return Ok(()),
        Some(Yaml::Map(section)) => section,
        Some(_) => {
            return Err(MalformedConfig {
                config_path: config_path.to_string_lossy().into_owned(),
                reason: "'validation' must be a mapping".to_string(),
            })
        }
    };
    check_severity_map(
        &config_path,
        yaml_get(section, "rules"),
        "validation.rules",
        &["error", "warning", "off"],
    )?;
    check_severity_map(
        &config_path,
        yaml_get(section, "types"),
        "validation.types",
        &["error", "warning"],
    )
}

// ---------------------------------------------------------------------------
// GateFinding / GateReport
// ---------------------------------------------------------------------------

#[derive(Debug)]
pub struct GateFinding {
    pub source: &'static str,
    pub code: String,
    pub severity: String,
    pub enforcement: &'static str,
    pub path: String,
    pub line: Option<i64>,
    pub message: String,
}

pub struct GateReport {
    pub directory: String,
    pub recursive: bool,
    pub findings: Vec<GateFinding>,
}

impl GateReport {
    pub fn blocking(&self) -> Vec<&GateFinding> {
        self.findings
            .iter()
            .filter(|f| f.enforcement == ENFORCEMENT_BLOCKING)
            .collect()
    }

    pub fn advisory(&self) -> Vec<&GateFinding> {
        self.findings
            .iter()
            .filter(|f| f.enforcement == ENFORCEMENT_ADVISORY)
            .collect()
    }

    /// True when nothing is blocking — advisory findings never fail the gate.
    pub fn ok(&self) -> bool {
        self.blocking().is_empty()
    }
}

// ---------------------------------------------------------------------------
// build_gate
// ---------------------------------------------------------------------------

/// One normalized validate finding: `(code, severity, path, line, message)`.
type RawFinding = (String, String, String, Option<i64>, String);

/// One normalized `(code, severity, path, line, message)` per validate
/// finding: per-file issues (line-anchored) then OKF findings (file-level).
fn validate_findings(result: &DirectoryValidation) -> Vec<RawFinding> {
    let mut out = Vec::new();
    for file in &result.files {
        for issue in &file.issues {
            out.push((
                issue.code.clone(),
                issue.severity.to_string(),
                file.path.clone(),
                issue.line,
                issue.message.clone(),
            ));
        }
    }
    if let Some(okf) = &result.okf {
        for finding in &okf.findings {
            out.push((
                finding.code.clone(),
                finding.severity.clone(),
                finding.path.clone(),
                None,
                finding.message.clone(),
            ));
        }
    }
    out
}

/// `build_gate(directory, recursive)` — run validation, relationships, and
/// review, then enforce the corpus policy. Findings the policy turns `off`
/// are dropped; the rest sort by `(path, line or 0, source, code, message)`.
pub fn build_gate(directory: &str, recursive: bool) -> Result<GateReport, MalformedConfig> {
    // The oracle raises from load_enforcement_policy first, then
    // load_overrides — mirror that order so a doubly-malformed config
    // reports the enforcement error.
    let policy = load_enforcement_policy(directory)?;
    check_overrides(directory)?;

    let validation = validate_directory(directory, recursive);
    let relationships: RelationshipValidation = validate_relationships(directory, recursive);
    let items = corpus_items(directory, recursive);
    let portfolio = portfolio_from_corpus(directory, &items, recursive);
    let review: ReviewReport = review_from_portfolio(directory, portfolio, recursive);

    let mut findings: Vec<GateFinding> = Vec::new();
    let mut add = |source: &'static str,
                   code: String,
                   severity: String,
                   path: String,
                   line: Option<i64>,
                   message: String,
                   default: &'static str| {
        if let Some(enforcement) = policy.classify(&code, default) {
            findings.push(GateFinding {
                source,
                code,
                severity,
                enforcement,
                path,
                line,
                message,
            });
        }
    };

    // Validate: an "error" fails validation, so it is blocking by default;
    // warnings and OKF info findings are advisory.
    for (code, severity, path, line, message) in validate_findings(&validation) {
        let default = if severity == "error" {
            ENFORCEMENT_BLOCKING
        } else {
            ENFORCEMENT_ADVISORY
        };
        add(SOURCE_VALIDATE, code, severity, path, line, message, default);
    }

    // Relationships: every issue fails `--validate` today, so blocking by
    // default. Message and (percent-encoded) path come from the shared SARIF
    // result builder so the two surfaces can never drift.
    for issue in &relationships.issues {
        let (message, uri) = relationship_sarif_parts(issue);
        let severity = relationship_severity(&issue.code).to_string();
        add(
            SOURCE_RELATIONSHIPS,
            issue.code.clone(),
            severity,
            uri,
            None,
            message,
            ENFORCEMENT_BLOCKING,
        );
    }

    // Review: priority 1-2 findings fail review today, so they are blocking
    // by default; advisory priorities (3+) are advisory.
    for issue in &review.issues {
        let message = if issue.action.is_empty() {
            issue.message.clone()
        } else {
            format!("{} \u{2014} {}", issue.message, issue.action)
        };
        let default = if issue.priority <= PRIORITY_BROKEN_RELATIONSHIP {
            ENFORCEMENT_BLOCKING
        } else {
            ENFORCEMENT_ADVISORY
        };
        add(
            SOURCE_REVIEW,
            issue.code.clone(),
            issue.severity.clone(),
            issue.path.clone(),
            None,
            message,
            default,
        );
    }

    findings.sort_by(|a, b| {
        a.path
            .cmp(&b.path)
            .then(a.line.unwrap_or(0).cmp(&b.line.unwrap_or(0)))
            .then(a.source.cmp(b.source))
            .then(a.code.cmp(&b.code))
            .then(a.message.cmp(&b.message))
    });
    Ok(GateReport {
        directory: directory.to_string(),
        recursive,
        findings,
    })
}
